### PR TITLE
Change Instance/Binding Parameters into a runtime.RawExtension. 

### PR DESCRIPTION
**Pull request by vaikas-google**
Unmarshal as yaml and populate outgoing requests with it

### CODE REVIEW

File: pkg/apis/servicecatalog/types.go [hunk 1]
```diff
@@ -197,7 +196,9 @@ type InstanceSpec struct {
 	// ServicePlanName is the reference to the ServicePlan for this instance.
 	PlanName string `json:"planName"`
 
-	Parameters map[string]runtime.Object
+	// Parameters is a YAML representation of the properties to be
+	// passed to the underlying broker.
+	Parameters string `json:"parameters"`
```
Comment (MHBauer) [inline pkg/apis/servicecatalog/types.go]: I don't think these need json annotations.

Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: Morgan is correct - these aren't needed on the internal types

**Comment by pmorie** [comment]
Per our previous discussions, I think the convention that we have for fields like this is that they should be `runtime.RawExtension` in the versioned API and `runtime.Object` in the internal API.

However, that said, I am not finding anything about this in the API conventions document:

https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md

I've created kubernetes/community#357 to document the convention and rationale behind it.

cc @kubernetes/api-reviewers

**Comment by pmorie** [comment]
@vaikas-google I think it would be cool to have a fuzzer that chooses between 2 or 3 pre-selected JSON blobs to ensure that the Instance/Bindings roundtrip through conversions correctly.

File: pkg/apis/servicecatalog/types.go [hunk 2]
```diff
@@ -278,7 +279,9 @@ type BindingSpec struct {
 	// should be injected with the results of the binding.  Immutable.
 	AppLabelSelector metav1.LabelSelector
 
-	Parameters map[string]runtime.Object
+	// Parameters is a YAML representation of the properties to be
+	// passed to the underlying broker.
+	Parameters string `json:"parameters"`
```
Comment (arschles) [inline pkg/apis/servicecatalog/types.go]: provision parameters are a JSON object, so this should be `map[interface{}]interface{}` to add specificity

Comment (arschles) [inline pkg/apis/servicecatalog/types.go]: also note that this can still be properly round-tripped by the yaml coder

Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: `interface{}` is not handled in the API machinery.  Fields that hold raw serialized JSON/YAML data should be `runtime.RawExtension` in the versioned API and `runtime.Object` in the internal API.

Comment (pmorie) [inline pkg/apis/servicecatalog/types.go]: See also: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/runtime/types.go#L47

Comment (arschles) [inline pkg/apis/servicecatalog/types.go]: ok

**Comment by arschles** [comment]
@vaikas-google what is the status of this PR?

**Comment by vaikas-google** [comment]
I'm on vacation, hence the long response time :) I'll try to find some time tomorrow to take a look at it :)
I can fix the one actionable comment by Morgan, but seems like there's still an outstanding request by Paul to figure out what type to use? Is that the only actionable comment, or did I miss something else?

**Comment by arschles** [comment]
@vaikas-google no actionable stuff for me, so I've approved the review. let me know when you're ready for review again (ping me in slack) and I'll re-review

File: test/integration/clientset_test.go [hunk 3]
```diff
@@ -352,6 +405,34 @@ func TestBindingClient(t *testing.T) {
...
+	}
+	if parameters.Foo != "bar" {
+		t.Fatal("Didn't get back 'bar' value for key 'foo' was %+v", parameters)
+	}
+	if len(parameters.Baz) != 2 {
+		t.Fatal("Didn't get back two values for 'baz' array in parameters was %+v", parameters)
+	}
+	foundFirst := false
+	foundSecond := false
+	for _, val := range parameters.Baz {
```
Comment (MHBauer) [inline test/integration/clientset_test.go]: unstructured slice so we have to search for each, right?

File: pkg/controller/handler.go [hunk 4]
```diff
@@ -93,12 +93,19 @@ func (h *handler) createServiceInstance(in *servicecatalog.Instance) error {
 
 	client := h.newClientFunc(broker.Name, broker.Spec.URL, authUsername, authPassword)
 
-	// TODO: uncomment parameters line once parameters types are refactored.
-	// Make the request to instantiate.
+	parameters := make(map[string]interface{})
+	if len(in.Spec.Parameters.Raw) > 0 {
+		err = yaml.Unmarshal([]byte(in.Spec.Parameters.Raw), &parameters)
```
Comment (MHBauer) [inline pkg/controller/handler.go]: The unmarshal here and below are to change from bytes (of text) to a struct. Why a yaml unmarshal vs the json unmarshal?

**Comment by MHBauer** [comment]
from what I can tell, lgtm.

Please squish the first two commits and then split out the generated code into a separate commit. The third commit for the fuzz testing is good as is.

File: pkg/controller/handler.go (hunk 4 above)
Comment (pmorie) [inline pkg/controller/handler.go]: It makes sense to me to support yaml and json in our api.

Comment (MHBauer) [inline pkg/controller/handler.go]: So it is doing the right thing? I don't understand.

**Comment by vaikas-google** [comment]
@MHBauer I think I (with help from @martinmaly) was able to get the commits into the shape you want. PTAL.

**Comment by MHBauer** [comment]
Thank you very much. It's a beat, lgtm.

File: pkg/controller/handler.go (hunk 4 above)
Comment (MHBauer) [inline pkg/controller/handler.go]: this is the whole 'json is yaml' thing, right?

**Pull request merged**